test/testlib/minikube_utilities.go

```diff
@@ -29,21 +30,86 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+/** Lists of the teardown and diagnostic teardown funcs */
 var teardownLists = make(map[string][]func())
+var diagnosticTeardownLists = make(map[string][]func())
+
+/** Exported var - initialised from the EnvVar, but can be reset in code if desired */
+var AlwaysRunDiagnosticTeardowns = strings.EqualFold(os.Getenv(ALWAYS_RUN_DIAGNOSTIC_TEARDOWNS), "true")
 
 /**
- * add a teardown function to the named list - for later execution
+ * add a teardown function to the named list - for deferred execution.
+ *
+ * The teardown functions are called in reverse order of insertion, by a call to Teardown(name).
+ *
+ * The typical idiom is:
+ * <pre>
+ *   testlib.AddTeardown("DATABASE", func() { ...})
+ *   // possibly more testlib.AddTeardown("DATABASE", func() { ... })
+ *   defer testlib.Teardown("DATABASE")
+ * <pre>
  */
 func AddTeardown(name string, teardownFunc func()) {
 	teardownLists[name] = append(teardownLists[name], teardownFunc)
 }
 
+/**
+ * add a diagnostic teardown func to be called before any other teardowns in the named list - to aid diagnostics/debugging.
+ * This allows a diagnostic teardown to do such things as:
+ * <ul>
+ *   <li>Generate logging and debug information immediately prior to resurce teardown
+ *   <li>call time.Sleep() to allow inspection and/or debugging of the exit state before teardown.
+ * <ul>
+ *
+ * NOTE: it is generally undesirable to add multiple diagnostic teardowns that sleep - so it would usually be best to
+ * add any Sleep() debug teardown to the innermost teardown list.
+ * Nonetheless, there are use-cases where multiple Sleep() teardowns are useful - to allow inspecting different
+ * intermediate states.
+ */
+func AddDiagnosticTeardown(name string, condition interface{}, teardownFunc func()) {
+	tdfunc := func() {
+		shouldIdoIt := AlwaysRunDiagnosticTeardowns
+
+		if !shouldIdoIt {
+			switch c := condition.(type) {
+			case *testing.T:
+				shouldIdoIt = c.Failed()
+
+			case func() bool:
+				shouldIdoIt = c()
+
+			case bool:
+				shouldIdoIt = c
+
+			default:
+				shouldIdoIt = c != nil
+			}
+		}
+
+		if shouldIdoIt {
+			teardownFunc()
```

Comment:
Why we execute the diagnostic function here when it is executed as part of the `Teardown()`? Maybe I'm overlooking something so a quick usage example will help.

Reply:
This call to teadownFunc is in a func() (tdfunc) that is being created in this context so that it has access to the 'condition'  param in its closure.

The tdfunc is then called in teardown - and tdfunc contains the conditional logic to decide whether the original diagnostic teardownFunc is actually called or not.

Reply:
@sivanov-nuodb 
> I'm a bit distracted by 1000+ changes in vendor files. Should these be part of the review?

The changes in 'vendor' are the result of my rebasing to master, which includes a change that removes the 'vendor' dir entirely.
By doing a rebase, I was trying to avoid those changes in master showing up in this subsequent PR.

Reply:
There are a few short usage examples in the unit tests:
test/minikube/minikube_test_test.go

Reply:
I also missed `tdfunc()`. I get how it works now.

Reply:
I will add more comments in future...

Reply:
I have added some additional comments and 2 usage examples in the PR description.

Reply:
Ah, it makes sense ... You create a closure/wrapper of the user-defined teardown function. Thanks for the explanation!